### ISSUE TITLE
Formatter should handle asserts in initializer list.

### ISSUE THREAD

**Issue by lrhn** [dart-lang/dart_style#522]
See dart-lang/sdk#27141 and dart-lang/sdk#27142 for context.

When the analyzer supports the syntax for assert initializers, the formatter needs to handle it gracefully.

**Comment by sethladd** [comment]
Hi! This issue is open, but I think we addressed this. Should we close?

cc @munificent

**Comment by sethladd** [comment]
Or, maybe not:

```
~/tmp $ dartfmt const_asserts.dart 
Could not format because the source could not be parsed:

line 6, column 7 of const_asserts.dart: Expected an identifier.
    : assert(children != null), assert(!children.any((child) => child == null));
      ^^^^^^
line 6, column 7 of const_asserts.dart: Expected an assignment after the field name.
    : assert(children != null), assert(!children.any((child) => child == null));
      ^^^^^^
line 6, column 7 of const_asserts.dart: Expected an identifier.
    : assert(children != null), assert(!children.any((child) => child == null));
      ^^^^^^
line 6, column 7 of const_asserts.dart: A function body must be provided.
    : assert(children != null), assert(!children.any((child) => child == null));
      ^^^^^^
line 6, column 7 of const_asserts.dart: Expected a class member.
    : assert(children != null), assert(!children.any((child) => child == null));
      ^^^^^^
line 6, column 7 of const_asserts.dart: Unexpected text 'assert'.
    : assert(children != null), assert(!children.any((child) => child == null));
      ^^^^^^
line 6, column 13 of const_asserts.dart: Expected a class member.
    : assert(children != null), assert(!children.any((child) => child == null));
            ^
line 6, column 13 of const_asserts.dart: Unexpected text '('.
    : assert(children != null), assert(!children.any((child) => child == null));
            ^
line 6, column 23 of const_asserts.dart: Expected a class member.
    : assert(children != null), assert(!children.any((child) => child == null));
                      ^^
line 6, column 23 of const_asserts.dart: Expected a class member.
    : assert(children != null), assert(!children.any((child) => child == null));
                      ^^
(23 more errors...)
~/tmp $ dartfmt --assert-initializers const_asserts.dart 
Could not find an option named "assert-initializers".

Usage: dartfmt [-n|-w] [files or directories...]

-h, --help                   Shows usage information.
    --version                Shows version information.
-l, --line-length            Wrap lines longer than this.
                             (defaults to "80")

-i, --indent                 Spaces of leading indentation.
                             (defaults to "0")

    --preserve               Selection to preserve, formatted as "start:length".
-n, --dry-run                Show which files would be modified but make no changes.
    --set-exit-if-changed    Return exit code 1 if there are any formatting changes.
-w, --overwrite              Overwrite input files with formatted output.
-m, --machine                Produce machine-readable JSON output.
    --profile                Display profile times after running.
    --follow-links           Follow links to files and directories.
                             If unset, links will be ignored.

-t, --transform              Unused flag for compability with the old formatter.
```

I'm using dartfmt 0.2.13, not sure if that's the latest?

**Comment by sethladd** [comment]
Ah, maybe we just need to do:

```
    // Parse it.
    var parser = new Parser(stringSource, errorListener);
    parser.enableAssertInitializer = true;
```

in dart_formatter.dart ?

**Comment by sethladd** [comment]
Tried the branch. Looks like it works for the simple case:

```
~/tmp $ dart ~/Code/dart_style/bin/format.dart const_asserts.dart 
class Foo {
  final List children;

  const Foo({this.children}) : assert(children != null);
}

main() {
  new Foo(children: null); // no failure

  const Foo(children: const []); // no failure

  const Foo(children: null); // failure
}
```

Thanks!

(no CLI argument required)

**Comment by sethladd** [comment]
I think this is done, yes?

**Comment by goderbauer** [comment]
I just wanted to setup the formatter for the flutter/plugins repo and the formatter is failing to format the following files because of this issue:
* https://github.com/flutter/plugins/blob/master/packages/firebase_database/lib/src/query.dart
* https://github.com/flutter/plugins/blob/master/packages/android_intent/lib/android_intent.dart

/cc @tvolkert

**Comment by kevmoo** [comment]
Hitting this a lot w/ https://github.com/dart-lang/pana – FYI

**Comment by dgrove** [comment]
@munificent any thoughts on when this could be addressed?

**Comment by munificent** [comment]
I'll get on it today.

**Comment by kevmoo** [comment]
Points for rolling this into the SDK 😁

**Comment by munificent** [comment]
Waiting until the changes from @nex3 are done first, then yes.

**Comment by goderbauer** [comment]
When is this change supposed to be available in the Dart SDK? The latest version 1.25.0-dev.4.0 doesn't seem to include it.

**Comment by munificent** [comment]
Soon. I was going to wait until those other changes are done, but that may take a while, so I'm going to go ahead and pull it in now.
